src/librustc_mir/interpret/eval_context.rs

```diff
@@ -546,57 +566,143 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tc
+            (Unwinding { in_cleanup: true }, true) => Unwinding { in_cleanup: true },
+            // We should never stop unwinding while in a cleanup block - cleanup
+            // blocks are generated by the compiler, and should never 'catch'
+            // an unwind
+            (Unwinding { in_cleanup: true }, false) => {
+                panic!("Attempted to stop unwinding during cleanup!")
+            },
+            // We want to keep unwinding - our cleanup state doesn't change,
+            // since we haven't hit a terminator with a cleanup block
+            (Unwinding { in_cleanup: false}, true) => Unwinding { in_cleanup: false },
+            // The machine has indicated that unwinding is now complete
+            (Unwinding { in_cleanup: false}, false) => NotUnwinding
+        };
+        info!("unwinding: state before {:?} from stack_pop {:?} after {:?}",
+              prev, stack_pop_info.unwinding, self.unwinding);
+
         // Abort early if we do not want to clean up: We also avoid validation in that case,
         // because this is CTFE and the final value will be thoroughly validated anyway.
         match frame.return_to_block {
-            StackPopCleanup::Goto(_) => {},
-            StackPopCleanup::None { cleanup } => {
+            StackPopCleanup::Goto{ .. } => {},
+            StackPopCleanup::None { cleanup, .. } => {
                 if !cleanup {
                     assert!(self.stack.is_empty(), "only the topmost frame should ever be leaked");
                     // Leak the locals, skip validation.
-                    return Ok(());
+                    return Ok(frame.return_to_block);
                 }
             }
         }
         // Deallocate all locals that are backed by an allocation.
         for local in frame.locals {
             self.deallocate_local(local.value)?;
         }
-        // Validate the return value. Do this after deallocating so that we catch dangling
-        // references.
-        if let Some(return_place) = frame.return_place {
-            if M::enforce_validity(self) {
-                // Data got changed, better make sure it matches the type!
-                // It is still possible that the return place held invalid data while
-                // the function is running, but that's okay because nobody could have
-                // accessed that same data from the "outside" to observe any broken
-                // invariant -- that is, unless a function somehow has a ptr to
-                // its return place... but the way MIR is currently generated, the
-                // return place is always a local and then this cannot happen.
-                self.validate_operand(
-                    self.place_to_op(return_place)?,
-                    vec![],
-                    None,
-                    /*const_mode*/false,
-                )?;
+
+
+        let do_normal_ret = self.unwinding != Unwinding { in_cleanup: false };
```

Comment:
So with `in_cleanup: true` we are *not* actually unwinding...?

Reply:
Yes - when `in_cleanup` is `true`, we're executing a cleanup block, which works exactly like normal execution (except for the fact that it eventually reaches a `Resume`).

Reply:
But then why is that in the `Unwinding` state if it's not unwinding? I don't get the point of this.

*All* of unwinding works exactly like normal execution. That's why I am baffled by all the extra complexity here. The only bit there is different is which block we jump to when popping a frame.

Reply:
Oh I get it, this is the part where we are unwinding but we are executing a destructor.